### PR TITLE
Add UI response towards selection event in profiler.

### PR DESCRIPTION
**Pull request by stellama0208**
- Change CPU slices' color at single-click selection, group slices based on process/thread ID.
 - Mark the selected item at single-click selection.
 - Mark multiple items at area selection triggered by SHIFT + mouse dragging .

### CODE REVIEW

**Comment by stellama0208** [comment]
The above force push is only to resolve merge conflict with PR#3292.

File: gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java [hunk 1]
```diff
@@ -137,15 +141,17 @@ public static Data empty(DataRequest req) {
...
     public final long[] ts;
     public final String[] names;
     public final double[][] values;
+    private final Set<Long> valueKeys = Sets.newHashSet();
 
     public Values(String name, Data data) {
       this.ts = data.ts;
       this.names = new String[] { name };
       this.values = new double[][] { data.values };
+      this.valueKeys.addAll(LongStream.of(ts).boxed().collect(Collectors.toList()));
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java]: `LongStream.of(ts).boxed().forEach(this.valueKeys::add)` - avoids creating and collecting to the throw-away list.

File: gapic/src/main/com/google/gapid/perfetto/models/Selection.java [hunk 2]
```diff
@@ -17,95 +17,140 @@
...
+        return firstSelection().buildUi(parent, state);
       } else {
         return new MultiSelectionView(parent, selections, state);
       }
     }
 
-    @Override
-    public void mark(State state) {
+    public <Key> Selection<Key> getSelection(Kind<Key> type) {
+      return selections.containsKey(type) ? (Selection<Key>) selections.get(type) :type.EmptySelection();
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/Selection.java]: Add the `@SupressWarnign` annotation here, since this should warn about the cast.

(Also small ws nit after the `:`)

File: gapic/src/main/com/google/gapid/perfetto/models/Selection.java [hunk 3]
```diff
@@ -17,95 +17,140 @@
...
   }
+
+  public static class Kind<Key> implements Comparable<Kind<?>>{
+    public int priority;
+
+    public Kind(int priority) {
+      this.priority = priority;
+    }
+
+    public Selection<Key> EmptySelection() {
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/Selection.java]: Functions should start with a lowercase function. Mirror `Collections.emptyList()`, which contains just a cast around a static final.

File: gapic/src/main/com/google/gapid/perfetto/models/Selection.java [hunk 4]
```diff
@@ -17,95 +17,140 @@
...
+    public int compareTo(Kind<?> other) {
+      return this.priority - other.priority;
+    }
+
+    public static final Kind<Slice.Key> Thread = new Kind<Slice.Key>(0);
+    public static final Kind<StateSlice.Key> ThreadState = new Kind<StateSlice.Key>(1);
+    public static final Kind<Long> Cpu = new Kind<Long>(2);
+    public static final Kind<Slice.Key> Gpu = new Kind<Slice.Key>(3);
+    public static final Kind<Long> Counter = new Kind<Long>(4);
+  }
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/Selection.java]: move these to the top of this class

File: gapic/src/main/com/google/gapid/perfetto/models/SliceTrack.java [hunk 5]
```diff
@@ -234,42 +235,49 @@ public Data(DataRequest request, long[] ids, long[] starts, long[] ends, int[] d
     }
   }
 
-  public static abstract class Slice implements Selection {
+  public static abstract class Slice implements Selection<Slice.Key> {
     public final long time;
     public final long dur;
     public final String category;
     public final String name;
+    public final long depth;
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/SliceTrack.java]: `int depth` - we've been using int everywhere else for depth.

File: gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java [hunk 6]
```diff
@@ -19,22 +19,26 @@
 import static com.google.gapid.widgets.Widgets.createStandardTabItem;
 
 import com.google.gapid.perfetto.models.Selection;
+import com.google.gapid.perfetto.models.Selection.Kind;
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java]: Refer to it as `Selection.Kind`

File: gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java [hunk 7]
```diff
@@ -19,22 +19,26 @@
...
 import org.eclipse.swt.widgets.TabFolder;
 
+import java.util.NavigableMap;
+
 /**
  * Displays multiple different selections.
  */
 public class MultiSelectionView extends Composite {
-  public MultiSelectionView(Composite parent, Selection[] selections, State state) {
+  public MultiSelectionView(Composite parent, NavigableMap<Kind<?>, Selection<?>> selections,
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java]: Maybe just ask for an `Iterable<Selection<?>>` or `Collection<Selection<?>>`.

File: gapic/src/main/com/google/gapid/perfetto/views/SelectionView.java [hunk 8]
```diff
@@ -49,12 +49,14 @@ public void onDataChanged() {
...
-  public void onSelectionChanged(Selection selection) {
+  public void onSelectionChanged(Selection.MultiSelection selection) {
     for (Control c : group.getChildren()) {
       c.dispose();
     }
 
-    if (selection != null) {
+    if (selection == null) {
+      new Composite(group, SWT.NONE).requestLayout();
+    } else {
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/SelectionView.java]: why create an empty composite and ask for layout?

File: gapic/src/main/com/google/gapid/perfetto/views/State.java [hunk 9]
```diff
@@ -55,8 +55,10 @@
   private double maxScrollOffset = 0;
   private double nanosPerPx;
   private long resolution;
-  private Selection selection;
+  private Selection.MultiSelection selection;
   private final AtomicInteger lastSelectionUpdateId = new AtomicInteger(0);
+  private long selectedUpid = -1;   // For a selected CPU slice.
+  private long selectedUtid = -1;   // For a selected CPU slice.
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/State.java]: Just `private ThreadInfo selectedThread`. The `ThreadInfo` keeps track of the `utid` and `upid` for you.

File: gapic/src/main/com/google/gapid/perfetto/canvas/Panel.java [hunk 10]
```diff
@@ -26,6 +25,12 @@
     return Hover.NONE;
   }
 
+  @SuppressWarnings("unused")
+  /** Returns whether the screen should be redrawn. */
+  public default boolean onClick(Fonts.TextMeasurer m, double x, double y) {
+    return false;
+  }
+
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/canvas/Panel.java]: Please undo this. I see a lot of duplicated code due to this as we have to re-compute where the mouse is now that this is a panel-level method, rather than on the `Hover` object.

File: gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java [hunk 11]
```diff
@@ -159,6 +168,14 @@ private void renderSlices(RenderContext ctx, CpuTrack.Data data, double h) {
       }
     }
 
+    // Draw bounding rectangles after all the slices are rendered, so that the border is on the top.
+    ctx.setForegroundColor(SWT.COLOR_BLACK);
+    for (int index : visibleSelected) {
+      double rectStart = state.timeToPx(data.starts[index]);
+      double rectWidth = Math.max(1, state.timeToPx(data.ends[index]) - rectStart);
+      ctx.drawRect(rectStart, 0, rectWidth, h, 3);
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java]: make 3 a constant

File: gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java [hunk 12]
```diff
@@ -159,6 +168,14 @@ private void renderSlices(RenderContext ctx, CpuTrack.Data data, double h) {
       }
     }
 
+    // Draw bounding rectangles after all the slices are rendered, so that the border is on the top.
+    ctx.setForegroundColor(SWT.COLOR_BLACK);
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java]: make the color a style constant. Black doesn't really work in dark mode.

File: gapic/src/main/com/google/gapid/perfetto/views/RootPanel.java [hunk 13]
```diff
@@ -311,8 +311,8 @@ protected void finishSelection() {
 
     Selection.CombiningBuilder builder = new Selection.CombiningBuilder();
     visit(Visitor.of(Selectable.class, (s, a) -> s.computeSelection(builder, a, ts)), selection);
-    selection = Area.NONE;
     state.setSelection(builder.build());
+    selection = Area.NONE;
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/views/RootPanel.java]: undo this change

File: gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java (hunk 7 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/MultiSelectionView.java]: If I understand it right: I addressed this comment by changing to a more abstract type Map<Kind<?>, Selection<?>>.

File: gapic/src/main/com/google/gapid/perfetto/views/SelectionView.java (hunk 8 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/SelectionView.java]: Sorry, was trying to clear the info box at a de-selection, but didn't notice the dispose() code above already does the clearing duty. Have reverted it back in the new commit.

File: gapic/src/main/com/google/gapid/perfetto/canvas/Panel.java (hunk 10 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/canvas/Panel.java]: Was trying to solve the de-selection problem in this way, but yeah it indeed included a lot of redundant code and the disadvantages outweighs the advantages. Have just reverted this part back, and found another way to solve the de-selection problem in RootPanel! Thanks Pascal!

File: gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java (hunk 12 above)
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/CpuPanel.java]: Will change the color for the bounding boxes in a new PR following.

**Comment by stellama0208** [comment]
Hi Pascal, the above comments are addressed in a new commit, it's ready for a new round of review, thanks!

File: gapic/src/main/com/google/gapid/perfetto/models/Selection.java [hunk 14]
```diff
@@ -17,95 +17,149 @@
...
 
-    public MultiSelection(Selection[] selections) {
-      this.selections = selections;
-    }
+  @SuppressWarnings("unchecked")
+  public static <K> Selection<K> emptySelection() {
+      return EMPTY_SELECTION;
+  }
 
+  public static class EmptySelection<K> implements Selection<K> {
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/perfetto/models/Selection.java]: private

Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/models/Selection.java]: For usage it should be private, but unfortunately changing it to private here will trigger grammar error "The interface member type EmptySelection can only be public". :(
